### ISSUE TITLE
Chamber 15 - visual errors

### ISSUE THREAD

**Issue by hackgrid** [mwpenny/portal64-still-alive#78]
#### Version of the game
 23464eea7c09c5bd3e2bedf1ac3cd4ac6576aeec

#### Original Hardware or Emulator?
Ares v139

#### What is the bug?

Great work on chapter 15, it looks and feels great, thank you! :-)

In chamber 15, last room, on the right button room:

If you stand in the corner, everything looks ok (but you can't see the flying ball through the window - only some split second while it enters the blue portal).

![portal 2024-08-03 19-08-06](https://github.com/user-attachments/assets/9d9a0297-0c94-4fe2-9b12-0543aa553439)

If you look a tiny bit further left, the button disappears:

![portal 2024-08-03 19-08-10](https://github.com/user-attachments/assets/5e567d8c-91ed-48d6-aa66-4dca2aa9fa41)

A tiny bit even further disappears the room :D

![portal 2024-08-03 19-08-15](https://github.com/user-attachments/assets/458c462b-186a-4687-aa0b-96f8f061aa8e)

After entering the last room, I experiences some visual erros there in general:

![portal 2024-08-03 19-18-54](https://github.com/user-attachments/assets/a1a9e460-c192-4645-9c41-c55491d25ea5)

After exiting the blue portal:
![portal 2024-08-03 19-18-30](https://github.com/user-attachments/assets/9561cee9-8053-4217-a127-913e8e451ac0)

![portal 2024-08-03 19-18-38](https://github.com/user-attachments/assets/a41930aa-c26e-4ba1-92e9-219c7df2b52a)

**Comment by mwpenny** [comment]
Thanks for testing things out and logging this! Glad you're enjoying.

What's happening here is that the game is trying to render too many objects. There's an internal limit on how many objects can be drawn each frame. Any found after the limit is reached will be ignored - hence the disappearing.

So why are so many objects being processed? The game thinks you can see into more of the earlier rooms than you're actually able to. A room is considered visible as long as one of its [`@doorway`](https://github.com/mwpenny/portal64-still-alive/blob/master/documentation/levels/level_objects/doorway.md)s is in the field of view, and obstructions are disregarded. So looking in their general direction is enough. Chamber 15 is a pretty long level and earlier rooms are evaluated first, so looking toward the beginning of the level causes the internal buffer to fill up from earlier rooms and hide parts of the last room.

This used to be much easier to hit. I improved it in 220dd58 by narrowing the view frustum to doorway bounds when doing recursive checks, but it's still possible to run into the issue as you can see (albeit less often). Another spot is in the second room: the camera and ball can be made to disappear when entering the room and looking toward the next room.

**Comment by mwpenny** [comment]
I think the best path forward is optimization at the macro level where there are likely more gains to be had. We can precalculate at build time which rooms are _potentially_ visible from a given room. That way all other rooms can be quickly ignored at runtime (at the expense of extra level export logic). Something similar to this https://youtu.be/IfCRHSIg6zo.

The view frustum based checks are imperfect but performant, and effective in most cases. Combining both should avoid the problem while keeping things fast and relatively simple.

**Comment by hackgrid** [comment]
> Thanks for testing things out and logging this! Glad you're enjoying.
> 
> What's happening here is that the game is trying to render too many objects. There's an internal limit on how many objects can be drawn each frame. Any found after the limit is reached will be ignored - hence the disappearing.
> 
> So why are so many objects being processed? The game thinks you can see into more of the earlier rooms than you're actually able to. A room is considered visible as long as one of its [`@doorway`](https://github.com/mwpenny/portal64-still-alive/blob/master/documentation/levels/level_objects/doorway.md)s is in the field of view, and obstructions are disregarded. So looking in their general direction is enough. Chamber 15 is a pretty long level and earlier rooms are evaluated first, so looking toward the beginning of the level causes the internal buffer to fill up from earlier rooms and hide parts of the last room.
> 
> This used to be much easier to hit. I improved it in [220dd58](https://github.com/mwpenny/portal64-still-alive/commit/220dd585b435f87b134658da1e47220d3223ca53) by narrowing the view frustum to doorway bounds when doing recursive checks, but it's still possible to run into the issue as you can see (albeit less often). Another spot is in the second room: the camera and ball can be made to disappear when entering the room and looking toward the next room.

Cool, thanks for your detailed response! :-)

I do not know anything about 3D programming, but should it not be possible to limit the processing only to objects in sight?

Since chamber 15 is so big, can't processing be limited to only the last room?
You can't see anything from the previous rooms?

Maybe you could (only for last room in chamber 15) set a "only process objects inside this room"-flag and ignore processing for all objects inside other rooms? Since the last room is special and you can't really see much of the previous rooms because of the narrow and twisted moving-platform-corridor?

Or could you split the chamber into like chamber15_part1 and _part2?
You could duplicate the corridor with the moving platforms in both parts, and load vice versa if you are past a certain point in the corridor?

Loading a new level is pretty fast and it should not lag, right?

Sorry if all that stuff was obvious and dumb :D

**Comment by hackgrid** [comment]
> I think the best path forward is optimization at the macro level where there are likely more gains to be had. We can precalculate at build time which rooms are _potentially_ visible from a given room. That way all other rooms can be quickly ignored at runtime (at the expense of extra level export logic). Something similar to this https://youtu.be/IfCRHSIg6zo.
> 
> The view frustum based checks are imperfect but performant, and effective in most cases. Combining both should avoid the problem while keeping things fast and relatively simple.

Seems cool! :-) Maybe you don't need to precalculate, since all portal chambers are known, we could manually set which rooms are visible from each other rooms, yes? :-) If only rooms 2 + 15 have these problems, maybe define this for them and ignore the other rooms, since they do not have this problem?

**Comment by mwpenny** [comment]
> should it not be possible to limit the processing only to objects in sight?

That's the idea behind the current code, but it doesn't check if there's something in the way occluding a given object. It just relies on direction and field of view. It's technically possible to change this, but it can be slow which is likely why James didn't do it (diminishing returns). This logic isn't just intended for keeping the number of objects low, but also improving performance.

So I'd rather try the fast method above first since it can quickly discard entire rooms, the let the imperfect-but-fast checks handle the doorways since truly invisible rooms will always be ignored at that point anyway.

> can't processing be limited to only the last room?
> [...]
> "only process objects inside this room"-flag
> [...]
> the last room is special and you can't really see much of the previous rooms because of the narrow and twisted moving-platform-corridor?

These are all possible, and the flag would be quite easy to add, but I'd prefer a general approach so we don't need to tag things manually when creating levels. Both to avoid mistakes and save on effort finding and marking problem areas. It shouldn't be too difficult, and the manual solution is always possible as a fallback :)

> Or could you split the chamber into like chamber15_part1 and _part2?

You're absolutely right that this would solve it. Loading a level is pretty fast but there's a brief hiccup. Players are also able to place portals before the hallway, so that would need to be handled. I'm confident it's solvable without this, but it will be needed for the escape sequence (just like the original game).

No dumb ideas! Brainstorming is helpful for considering everything before going forward, thanks!

**Comment by hackgrid** [comment]
Thanks for your insights! :) Looking forward to what you are cooking!

**Comment by mwpenny** [comment]
I've been sick and taking it easy and am now gradually coming back to this with a few experiments. Future chambers will be larger than 15 and so it may be worth it to do multiple layers of optimization eventually. Better to know now one way or the other.

* As expected, hard-coding the list of potentially visible rooms completely solves the problem, at the cost of some manual work.
* It's also solvable without new functionality by creating more logical rooms in between other rooms to block doorway line of sight.

  For example, given rooms A-B-C connected through single doorways, the game knows a player in A cannot see C unless they can see its doorway through B's doorway (excluding walls, etc.; and portals are handled in a separate pass). So in certain situations it's possible to organize rooms in a way that makes this impossible. It's more work when making the level, and some work would need to be done to fix some issues this introduces in chamber 15, but it's a nice fallback / good practice to follow.
* Processing rooms from nearest to farthest helps in some cases like the end of chamber 15, but only for static geometry (dynamic objects are organized differently).
* I implemented a hacky version of frustum culling through doorways. It helps keep the object count down, but not enough to matter right now. Something to keep in our back pockets down the line.

So I've validated the approach mentioned earlier in this thread and learned some more about others which could be useful to add on later. Now I just need to do the automatic detection part.

This whole thing is making me want to redo the culling system :P (e.g., using BSP trees). Might be worth it when doing the escape sequence.

**Comment by hackgrid** [comment]
Very cool, thanks for you hard work! Take care of yourself first!

**Comment by mwpenny** [comment]
@hackgrid after looking more into this, the game's data structures aren't really set up to answer the question "is this doorway visible" at build time without resorting to some hackiness.

I decided to go with your suggestion of leaving it as a manually-specified option. For a given room, we can now explicitly constrain the rooms that it is allowed to see. If not specified then the game behaves as before. See updated [room.md](https://github.com/mwpenny/portal64-still-alive/blob/master/documentation/levels/level_objects/room.md).

Since most of the time the visibility problem can be corrected by adding additional rooms and doorways, I don't think this will be needed often and so it's likely not worth it to polish too much here. Rooms couldn't be split for chamber 15 since it would mean animated objects would be cross boundaries, which is not allowed (there are ways to fix this but that's even more work). Also if my hunch is correct the underlying data structures will need to change in the future anyway :)

So maybe one day this can be removed in favour of something better, but it works quite well. Give it a try! I found and fixed another issue while I was at it which caused the crosshair to not work properly when looking through multiple doorways.
